Resp EVAL/EVAL_RO/EVALSHA/EVALSHA_RO/SCRIPT LOAD|EXISTS|FLUSH

tristantarrant opened the pull request:
Depends on #13190 

I will also add a section on scripting to the resp docs once #13198 is merged

Closes #13192

tristantarrant commented:
I need to exclude the lua stuff from the native server

Saulzi commented:
Does this add support for these resp commands?

I am looking at the feasibility of using infinispan as a drop in replacement for the Redis, unfortunately it will not work with the .NET client for redis (StackExchange.Redis) as Infinispan does not support the eval command, I am not sure which other commands may be required,

It does not look like the HotRod client for .NET has been updated in some time, 

Really what I am looking to do is use the Infinispan as a distributed cache in .NET (IDistributedCache) but I cannot find any recent / supported libraries to do this,

tristantarrant commented:
Yes, this adds those commands. I will use that client as additional validation that it  works and I'll update you.

rigazilla commented on server/resp/src/main/java/org/infinispan/server/resp/RespRequestHandler.java:
```diff
@@ -62,10 +62,16 @@ public CompletionStage<RespRequestHandler> myStage() {
     */
    public ResponseWriter writer() {
-      assert ctx.channel().eventLoop().inEventLoop() : "Buffer allocation should occur in event loop, it was " + Thread.currentThread().getName();
+      assert writer.isInternal() || ctx.channel().eventLoop().inEventLoop() : "Buffer allocation should occur in event loop, it was " + Thread.currentThread().getName();
```
I have test errors for writer==null

tristantarrant replied:
Fixed

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaTaskEngine.java:
```diff
@@ -0,0 +1,380 @@
+            lua.push("ok");
+            lua.rawGet(-2);
+            if (lua.type(-1) == Lua.LuaType.STRING) {
+               String ok = lua.toString(-1).replaceAll("[\\r\\n]", " ");
+               handler.writer().string(ok);
```
These characters should be fine with the `string` method. It could be problematic for a simple string.

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/serialization/lua/LuaResponseWriter.java:
```diff
@@ -0,0 +1,221 @@
+   }
+
+   @Override
+   public void simpleString(CharSequence value) {
```
I guess these string, integer, and double methods needs to check for null.

tristantarrant replied:
I've added null checks to all methods

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/serialization/lua/LuaResponseWriter.java:
```diff
@@ -0,0 +1,221 @@
+
+   @Override
+   public void writeNumericPrefix(byte symbol, long number, int additionalWidth) {
+      throw new UnsupportedOperationException(CallerId.getCallerMethodName(1));
```
Should it be no-op? I see the method is called directly by some commands in the sorted set.

wburns replied:
Agreed, I don't feel comfortable with any of these methods throwing an exception tbh.

tristantarrant replied:
I've rewritten most of the commands doing this to use proper RESP structures, instead of doing things by hand. The only significant leftover is `EXEC`. It looks like I should rewrite it to collect all responses into an array and write it at the end

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/commands/scripting/eval/EVAL.java:
```diff
@@ -0,0 +1,65 @@
+public class EVAL extends RespCommand implements Resp3Command {
+
+   public EVAL() {
+      super(-3, 1, 1, 1);
```
Seems it should be `-3, 0, 0, 0`:

```text
350)  1) "eval"
      2) (integer) -3
      3) 1) noscript
         2) stale
         3) skip_monitor
         4) no_mandatory_keys
         5) movablekeys
      4) (integer) 0
      5) (integer) 0
      6) (integer) 0

```

tristantarrant replied:
Fixed

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/commands/scripting/eval/FLUSH.java:
```diff
@@ -0,0 +1,51 @@
+   private static final byte[] ASYNC_BYTES = "ASYNC".getBytes(StandardCharsets.US_ASCII);
+
+   protected FLUSH() {
+      super(2, 0, 0, 0);
```
It should be -2.

tristantarrant replied:
Fixed

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/commands/scripting/eval/KILL.java:
```diff
@@ -0,0 +1,34 @@
+ */
+public class KILL extends RespCommand implements Resp3Command {
+   protected KILL() {
+      super(-3, 0, 0, 0);
```
This one is 2 instead.

tristantarrant replied:
Fixed

jabolina reviewed (commented):
Looking great, just a few things.

rigazilla commented on server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaArray.java:
```diff
@@ -0,0 +1,111 @@
+      @Override
+      public V next() {
+         lua.rawGetI(index, ++i);
+         return null;
```
always returns null, is it ok?

tristantarrant replied:
I've added a comment clarifying the way it work

rigazilla commented on server/resp/src/main/java/org/infinispan/server/resp/commands/scripting/eval/DEBUG.java:
```diff
@@ -0,0 +1,34 @@
+ * SCRIPT DEBUG
+ *
+ * @see <a href="https://redis.io/commands/script-debug/">SCRIPT DEBUG</a>
+ * @since 15.1
```
```suggestion
 * @since 15.2
```

tristantarrant replied:
Fixed

rigazilla commented on server/resp/src/main/java/org/infinispan/server/resp/commands/scripting/eval/EVAL.java:
```diff
@@ -0,0 +1,65 @@
+      String script = ascii(arguments.get(0));
+      int numKeys = (int) ArgumentUtils.toLong(arguments.get(1));
+      String[] keys = new String[numKeys];
+      for (int i = 0; i < numKeys; i++) {
```
should we check the number of args?
```
127.0.0.1:6379> eval "return 0" 3 a
(error) ERR Number of keys can't be greater than number of args
```

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/RespServer.java:
```diff
@@ -179,4 +203,8 @@ public SegmentSlotRelation segmentSlotRelation() {
    }
+
+   public Random random() {
+      return random;
```
If we want to mirror the original behaviour, where there is only one thread, there should only be a single instance.

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaContext.java:
```diff
@@ -0,0 +1,455 @@
+      lua.setTable(-3);
+      lua.push("randomseed");
+      lua.push(l -> {
+         handler.respServer().random().setSeed(l.toInteger(1));
```
Yes, the seed is shared.

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaContextPool.java:
```diff
@@ -0,0 +1,80 @@
+               pool.poll().shutdown();
+            }
+         }
+      }, validationInterval, validationInterval, TimeUnit.SECONDS);
```
I've removed the executor completely now.

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaContext.java:
```diff
@@ -0,0 +1,455 @@
+         if (resp != 2 && resp != 3) {
+            l.error("RESP version must be 2 or 3.");
+         }
+         // As Infinispan doesn't handle RESP 2, should we throw an error here ?
```
Ok, I've removed the comment

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaContext.java:
```diff
@@ -0,0 +1,455 @@
+      try {
+         future.get(); // TODO: handle timeouts ?
+      } catch (Throwable t) {
+         Throwable cause = filterCause(t);
```
I've added a debug message

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaContext.java:
```diff
@@ -0,0 +1,455 @@
+      for (int i = -argc + 1; i < 0; i++) {
+         args.add(l.toString(i).getBytes(StandardCharsets.US_ASCII));
+      }
+      CompletableFuture<RespRequestHandler> future = ((Resp3Command) respCommand).perform(handler, ctx, args).toCompletableFuture();
```
I've applied the suggestion

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/commands/scripting/eval/EVAL.java:
```diff
@@ -0,0 +1,65 @@
+ * </p>
+ *
+ * @see <a href="https://redis.io/docs/latest/commands/eval/">EVAL</a>
+ * @since 15.1
```
Fixed

tristantarrant commented on build/configuration/pom.xml:
```diff
@@ -135,6 +135,8 @@
       <version.commons.compress>1.27.1</version.commons.compress>
       <version.console>15.2.0.Final</version.console>
       <version.fabric8.kubernetes-client>6.13.5</version.fabric8.kubernetes-client>
+      <version.fabric8.kubernetes-client>6.13.4</version.fabric8.kubernetes-client>
```
Done

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/RespCommand.java:
```diff
@@ -120,6 +117,33 @@ public static RespCommand fromByteBuf(ByteBuf buf, int commandLength) {
+                  break next;
+               }
+            }
+            return possible;
```
Done

tristantarrant commented:
@jabolina @wburns I am not going to fix all the issues in this PR (custom serializers and RESP2). I'll just fix the current failures

tristantarrant commented:
I've fixed all the custom serializers, except some edge cases (`COMMAND` which cannot be invoked in a Lua script, and the `CLUSTER SLOTS|SHARDS`) and I've added the skeleton for handling RESP2 (although attempting to use it will throw an error).

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaTaskEngine.java:
```diff
@@ -0,0 +1,380 @@
+            luaCtx.registerScript(script);
+            runScript(luaCtx, handler, ctx, script, keys, args, flags);
+         } finally {
+            luaCtx.unregisterScript(script);
```
Do we need to close the luaCtx if there was an exception? It won't be returned to the pool and the other stuff in `thenApplyAsync` won't run either.

tristantarrant replied:
I've reworked this

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaTaskEngine.java:
```diff
@@ -0,0 +1,380 @@
+         try {
+            script = scriptingManager.getScriptWithMetadata(scriptName(sha.toUpperCase()));
+         } catch (CacheException e) {
+            throw new RuntimeException("NOSCRIPT No matching script. Please use EVAL.");
```
Same here we need to make sure to handle the ctx better when exceptions occur.

tristantarrant replied:
I've reworked this

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaTaskEngine.java:
```diff
@@ -0,0 +1,380 @@
+         TransactionContext.startTransactionContext(ctx);
+         runScript(luaCtx.lua, script, keys, args);
+      } catch (Throwable t) {
+         throw new RuntimeException(t);
```
This RuntimeException seems extraneous. I also worry that if an exception does happen, nothing is propagated to the user as the handler isn't written to.

tristantarrant replied:
With the change in the callers, this should be handled properly now

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/serialization/bytebuf/ByteBufCollectionSerializer.java:
```diff
@@ -56,8 +56,8 @@ public void accept(Collection<?> objects, ByteBufPool alloc, SerializationHint.S
-         // Handle any collection which is not a set.
-         return object instanceof Collection<?> && !(object instanceof Set<?>);
+         // Handle any collection
+         return object instanceof Collection<?>;
```
Is this change needed? I am a bit surprised at the change as this deals with the ByteBuf serializer..

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/serialization/lua/LuaResponseWriter.java:
```diff
@@ -0,0 +1,285 @@
+      if (value == null) {
+         lua.pushNil();
+      } else {
+         lua.push(new String(value, StandardCharsets.US_ASCII));
```
I don't know if we can assume US_ASCII here. I believe this can contain any user value, so it can be any byte value.

tristantarrant replied:
I've changed it to use ISO8859-1 instead and added an explanation

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/serialization/lua/LuaResponseWriter.java:
```diff
@@ -0,0 +1,285 @@
+         for (T o : collection) {
+            lua.push(i);
+            serializer.accept(o, this);
+            lua.setTable(-3);
```
Have to admit all these magical numbers are not helpful. Maybe we should have these special values on an interface commenting what they mean?

tristantarrant replied:
nothing magical here: https://www.lua.org/manual/5.1/manual.html#lua_settable

wburns replied:
It is magical in the sense that we have to look it up somewhere else that isn't in the code. I know it has a defined purpose.

tristantarrant replied:
I have sprinkled a bunch of comments which should make things clearer

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/serialization/lua/LuaResponseWriter.java:
```diff
@@ -0,0 +1,285 @@
+
+   @Override
+   public void emptySet() {
+      throw new UnsupportedOperationException("emptySet"); // Only used by COMMAND which cannot be called from Lua.
```
Is it possible we add other commands that may support this though? I fear no one will update this and the command added won't be tested by a lua script we are testing. Maybe we should still implement or is it difficult to do so?

tristantarrant replied:
Probably, but that means creating tests just to exercise parts of the code that won't ever be executed at the moment

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/serialization/lua/LuaResponseWriter.java:
```diff
@@ -0,0 +1,285 @@
+
+   @Override
+   public <T> void set(Set<?> set, JavaObjectSerializer<T> serializer) {
+      throw new UnsupportedOperationException(CallerId.getCallerMethodName(1));
```
Same here, I don't know if I like the idea of leaving these little time bombs that most likely will never seen unless someone in prod or a random dev/qe dog foods and finds it later.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/serialization/lua/LuaResponseWriter.java:
```diff
@@ -0,0 +1,285 @@
+
+   @Override
+   public <T> void write(T object, JavaObjectSerializer<T> serializer) {
+      throw new UnsupportedOperationException(CallerId.getCallerMethodName(1));
```
Shouldn't this work though if you use the provided serializer?

tristantarrant replied:
The "hand-written" serializers won't work: the lua responses must adhere to well-known RESP structures.

wburns commented on server/tests/src/test/java/org/infinispan/server/footprint/FootprintIT.java:
```diff
@@ -30,8 +30,8 @@ public class FootprintIT {
-   private static final long DISK_USAGE_LOWER_BOUND = 69_000_000L;
-   private static final long DISK_USAGE_UPPER_BOUND = 74_000_000L;
+   private static final long DISK_USAGE_LOWER_BOUND = 73_000_000L;
+   private static final long DISK_USAGE_UPPER_BOUND = 75_000_000L;
```
Is it okay changing the range from 5 MB to 2MB?

tristantarrant replied:
Disk footprint changes very rarely, and, unfortunately, it decreases even more rarely :)

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/RespVersion.java:
```diff
@@ -0,0 +1,18 @@
+ *
+ * @since 15.2
+ */
+public enum RespVersion {
```
The intention is that we also support RESP2 https://github.com/infinispan/infinispan/issues/13986

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/commands/scripting/eval/DEBUG.java:
```diff
@@ -0,0 +1,34 @@
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx, List<byte[]> arguments) {
+      return handler.myStage();
```
I'm removint DEBUG and KILL, since we cannot implement them without https://github.com/gudzpoz/luajava/issues/227

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/Resp3Handler.java:
```diff
@@ -88,8 +92,7 @@ public BlockingManager getBlockingManager() {
          List<byte[]> arguments) {
-      if (type instanceof Resp3Command) {
-         Resp3Command resp3Command = (Resp3Command) type;
+      if (type instanceof Resp3Command resp3Command) {
```
`HELLO` is not.

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/commands/scripting/eval/KILL.java:
```diff
@@ -0,0 +1,34 @@
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx, List<byte[]> arguments) {
+      return handler.myStage();
```
I'm removing this.

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/commands/scripting/eval/EXISTS.java:
```diff
@@ -0,0 +1,52 @@
+      }
+      try {
+         return handler.getBlockingManager()
+               .supplyBlocking(() -> handler.respServer().luaEngine().scriptExists(shas), "script exists")
```
This is actually simpler to implement as is, and since this should be an infrequently invoked command, I'd rather keep it this way.

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaArray.java:
```diff
@@ -0,0 +1,112 @@
+   private final int index;
+
+   public LuaArray(Lua lua, int index) {
+      this(lua, index, lua.rawLength(index));
```
index is the position on the Lua stack.

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/RespCommand.java:
```diff
@@ -120,6 +117,25 @@ public static RespCommand fromByteBuf(ByteBuf buf, int commandLength) {
    }
 
+   public static RespCommand fromString(String s) {
+      byte[] buf = s.getBytes(StandardCharsets.US_ASCII);
```
I've changed this so that no allocation happens.

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/scripting/ScriptFlags.java:
```diff
@@ -0,0 +1,24 @@
+package org.infinispan.server.resp.scripting;
+
+public enum ScriptFlags {
+   NO_WRITES(1),
```
I've added comments for all flags

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/commands/scripting/eval/FLUSH.java:
```diff
@@ -0,0 +1,51 @@
+ */
+public class FLUSH extends RespCommand implements Resp3Command {
+   private static final byte[] SYNC_BYTES = "SYNC".getBytes(StandardCharsets.US_ASCII);
+   private static final byte[] ASYNC_BYTES = "ASYNC".getBytes(StandardCharsets.US_ASCII);
```
I've removed them

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/response/LCSResponse.java:
```diff
@@ -35,31 +40,37 @@ public void accept(LCSResponse res, ResponseWriter writer) {
 
-         // LCS client library for tests assume the keys in this order.
-         writer.writeNumericPrefix(RespConstants.MAP, 2);
+         Map<byte[], Object> map = new LinkedHashMap<>(2);
```
I've explained the reason

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaContextPool.java:
```diff
@@ -0,0 +1,67 @@
+   LuaContextPool(Supplier<LuaContext> supplier, int min, int max) {
+      this.supplier = supplier;
+      this.max = max;
+      pool = new ConcurrentLinkedQueue<>();
```
Done

tristantarrant commented on server/resp/src/main/java/org/infinispan/server/resp/scripting/LuaMap.java:
```diff
@@ -0,0 +1,88 @@
+ * A read-only map implementation that maps lua tables. Only size and iterations are supported.
+ */
+public class LuaMap<K, V> implements Map<K, V> {
+   static final Entry<Object, Object> EMPTY_ENTRY = Map.of(new Object(), new Object()).entrySet().iterator().next();
```
Done

tristantarrant commented:
@wburns (replying here because GitHub doesn't let me reply to your comment):

> Looking at LuaContext, I don't think this will work as all of the lua instances will share the same Random object

That's the intention. In Redis there is only one engine

tristantarrant commented:
@wburns I should have addressed all issues

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/serialization/lua/LuaResponseWriter.java:
```diff
@@ -0,0 +1,344 @@
+   }
+
+   @Override
+   public void arrayNext() {
```
Have to admit, I am not a fan of this method from the caller perspective as I feel we could know we are inside an array and do this. But it does make this class a lot less cluttered, so I will not push back on it.

wburns commented:
Integrated into main, thanks @tristantarrant !
